### PR TITLE
Safely restart jenkins service

### PR DESCRIPTION
**Pull request by surminus**
When we manage files on disk, we have to restart the Jenkins service if we make a change. Doing a service refresh does not take into account if any builds may be happening and will cause them to fail.

This commit adds a class which pulls the jenkins-cli.jar and issues a safe restart. Use this class to notify rather than the service itself.

### CODE REVIEW

File: modules/govuk_jenkins/manifests/safe_restart.pp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+# [*jenkins_user*]
+#   Which user to run the jar as.
+#
+class govuk_jenkins::safe_restart (
+  $jenkins_home = $::govuk_jenkins::jenkins_homedir,
+  $jenkins_user = $::govuk_jenkins::jenkins_user,
+) {
+  require ::govuk_jenkins
+
+  exec { 'Jenkins CLI tool':
```
Comment (rjw1) [inline modules/govuk_jenkins/manifests/safe_restart.pp]: We could use [puppet-curl](https://github.com/alphagov/puppet-curl) here

**Comment by alexmuller** [comment]
Wow this is weird/cool.

File: modules/govuk_jenkins/manifests/safe_restart.pp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    command => "/usr/bin/wget http://localhost:8080/nlpJars/jenkins-cli.jar -O ${jenkins_home}/jenkins-cli.jar",
+    user    => $jenkins_user,
+    group   => $jenkins_user,
+    unless  => "/bin/ls ${jenkins_home}/jenkins-cli.jar 2>/dev/null",
+  }
+
+  exec { 'Restart Jenkins':
+    command => "/usr/bin/java -jar ${jenkins_home}/jenkins-cli.jar -s http://localhost:8080/ -i ${jenkins_home}/.ssh/id_rsa",
```
Comment (surminus) [inline modules/govuk_jenkins/manifests/safe_restart.pp]: I just realised this command is wrong

File: modules/govuk_jenkins/manifests/safe_restart.pp (hunk 1 above)
Comment (surminus) [inline modules/govuk_jenkins/manifests/safe_restart.pp]: I think that would be nice, but it doesn't give the ability to specify the user/group that the file gets set as. I'm wondering if it needs to be run as Jenkins user at all, actually.

**Comment by surminus** [comment]
I've changed this a little to use puppet-curl, and to just instigate the task as root.

**Comment by alexmuller** [comment]
Does the `exec` resource need a `refreshonly => true` set? And then `notify => Exec['Restart Jenkins']` or something like that? Otherwise I'm a little worried the command might execute every Puppet run.

**Comment by surminus** [comment]
The `notify` in the File resource should only call when the file changes. I call the `Class` instead of an `Exec` to ensure that the jenkins-cli.jar exists. We could order this in a different way, perhaps with a require on the exec itself?

**Comment by alexmuller** [comment]
But I guess the `exec` will still execute every time because that's how it behaves, right? Unless it has an `onlyif`, `unless` or `refreshonly` or something like this.

**Comment by djsd123** [comment]
@surminus For reference https://docs.puppet.com/puppet/3.8/type.html#exec-attribute-refreshonly

**Comment by surminus** [comment]
So does `notify => Service['jenkins'],` behave differently to `notify => Class['Govuk_jenkins::Safe_restart'],`? 

That's what I don't understand, as I thought the notify would only call the class if the content changed on the File resource.

**Comment by surminus** [comment]
I've added `refreshonly => true` anyway as it doesn't hurt.

**Comment by surminus** [comment]
Thanks for the patience everyone (and for @deanwilson explaining), I misunderstood and got thrown by how the `Service` resource worked.

**Comment by surminus** [comment]
@deanwilson updated so that it hardcodes the home directory.

**Pull request merged**